File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/voicechat/SvcPacketSender.java
package crabcraft.net.crabUtilities.voicechat;

import crabcraft.net.crabUtilities.CrabUtilities;
import org.bukkit.entity.Player;

import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.UUID;

/**
 * Sends fake SVC {@code PlayerStatePacket} and {@code RemovePlayerStatePacket}
 * messages to local clients, making cross-server group members appear in
 * the SVC group GUI.
 *
 * <p>We don't have access to SVC's internal {@code PlayerState} class so
 * we replicate its wire format manually (verified against
 * {@code common/.../PlayerState.java:80-104}):
 * <pre>
 *   boolean disabled
 *   boolean disconnected
 *   UUID    uuid           (16 bytes, big-endian high then low)
 *   String  name           (VarInt length + UTF-8 bytes)
 *   boolean hasGroup
 *   [if hasGroup] UUID group
 * </pre>
 *
 * <p>Channels: {@code voicechat:state} for state updates, {@code
 * voicechat:remove_state} for removals. Bukkit requires the plugin to
 * register these as outgoing channels before {@code sendPluginMessage}
 * works — done in {@link CrabUtilities#onEnable()}.
 */
class SvcPacketSender {

    static final String STATE_CHANNEL = "voicechat:state";
    static final String REMOVE_STATE_CHANNEL = "voicechat:remove_state";

    private final CrabUtilities plugin;

    SvcPacketSender(CrabUtilities plugin) {
        this.plugin = plugin;
    }

    /** Send a PlayerStatePacket to a single recipient. */
    void sendState(Player recipient, UUID playerUuid, String playerName, UUID groupId) {
        byte[] bytes = encodeState(playerUuid, playerName, groupId);
        if (bytes == null) return;
        try {
            recipient.sendPluginMessage(plugin, STATE_CHANNEL, bytes);
        } catch (Exception e) {
            plugin.getLogger().fine("Failed to send PlayerStatePacket to "
                    + recipient.getName() + ": " + e.getMessage());
        }
    }

    /** Send a RemovePlayerStatePacket to a single recipient. */
    void sendRemove(Player recipient, UUID playerUuid) {
        byte[] bytes = encodeRemove(playerUuid);
        try {
            recipient.sendPluginMessage(plugin, REMOVE_STATE_CHANNEL, bytes);
        } catch (Exception e) {
            plugin.getLogger().fine("Failed to send RemovePlayerStatePacket to "
                    + recipient.getName() + ": " + e.getMessage());
        }
    }

    private static byte[] encodeState(UUID playerUuid, String playerName, UUID groupId) {
        try {
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            DataOutputStream dout = new DataOutputStream(out);
            dout.writeBoolean(false);  // disabled — not propagated cross-server
            dout.writeBoolean(false);  // disconnected — not propagated cross-server
            dout.writeLong(playerUuid.getMostSignificantBits());
            dout.writeLong(playerUuid.getLeastSignificantBits());

            byte[] nameBytes = (playerName == null ? "" : playerName)
                    .getBytes(StandardCharsets.UTF_8);
            VarInt.write(out, nameBytes.length);
            out.write(nameBytes);

            dout.writeBoolean(groupId != null);
            if (groupId != null) {
                dout.writeLong(groupId.getMostSignificantBits());
                dout.writeLong(groupId.getLeastSignificantBits());
            }
            return out.toByteArray();
        } catch (IOException e) {
            return null;
        }
    }

    private static byte[] encodeRemove(UUID playerUuid) {
        ByteArrayOutputStream out = new ByteArrayOutputStream(16);
        try {
            DataOutputStream dout = new DataOutputStream(out);
            dout.writeLong(playerUuid.getMostSignificantBits());
            dout.writeLong(playerUuid.getLeastSignificantBits());
        } catch (IOException ignored) {
            // ByteArrayOutputStream never throws
        }
        return out.toByteArray();
    }
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/voicechat/SvcPacketSender.kt
package crabcraft.net.crabUtilities.voicechat

import crabcraft.net.crabUtilities.CrabUtilities
import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.UUID
import org.bukkit.entity.Player

/**
 * Sends SVC PlayerStatePacket and RemovePlayerStatePacket messages to local clients, making cross-server group members
 * appear in the SVC group GUI.
 *
 * The state wire format is: disabled boolean, disconnected boolean, UUID (big-endian high then low), name (VarInt byte
 * length + UTF-8), hasGroup boolean, optional group UUID. CrabUtilities registers both outgoing plugin channels during
 * enable.
 */
open class SvcPacketSender(private val plugin: CrabUtilities?) {
    /** Send a PlayerStatePacket to a single recipient. */
    open fun sendState(recipient: Player, playerUuid: UUID, playerName: String?, groupId: UUID?) {
        val bytes = encodeState(playerUuid, playerName, groupId) ?: return
        try {
            recipient.sendPluginMessage(plugin!!, STATE_CHANNEL, bytes)
        } catch (e: Exception) {
            plugin!!.getLogger().fine("Failed to send PlayerStatePacket to ${recipient.getName()}: ${e.message}")
        }
    }

    /** Send a RemovePlayerStatePacket to a single recipient. */
    open fun sendRemove(recipient: Player, playerUuid: UUID) {
        val bytes = encodeRemove(playerUuid)
        try {
            recipient.sendPluginMessage(plugin!!, REMOVE_STATE_CHANNEL, bytes)
        } catch (e: Exception) {
            plugin!!.getLogger().fine("Failed to send RemovePlayerStatePacket to ${recipient.getName()}: ${e.message}")
        }
    }

    companion object {
        const val STATE_CHANNEL = "voicechat:state"
        const val REMOVE_STATE_CHANNEL = "voicechat:remove_state"

        private fun encodeState(playerUuid: UUID, playerName: String?, groupId: UUID?): ByteArray? {
            return try {
                val out = ByteArrayOutputStream()
                val data = DataOutputStream(out)
                data.writeBoolean(false) // disabled — not propagated cross-server
                data.writeBoolean(false) // disconnected — not propagated cross-server
                data.writeLong(playerUuid.mostSignificantBits)
                data.writeLong(playerUuid.leastSignificantBits)
                val nameBytes = (playerName ?: "").toByteArray(StandardCharsets.UTF_8)
                VarInt.write(out, nameBytes.size)
                out.write(nameBytes)
                data.writeBoolean(groupId != null)
                if (groupId != null) {
                    data.writeLong(groupId.mostSignificantBits)
                    data.writeLong(groupId.leastSignificantBits)
                }
                out.toByteArray()
            } catch (_: IOException) {
                null
            }
        }

        private fun encodeRemove(playerUuid: UUID): ByteArray {
            val out = ByteArrayOutputStream(16)
            try {
                val data = DataOutputStream(out)
                data.writeLong(playerUuid.mostSignificantBits)
                data.writeLong(playerUuid.leastSignificantBits)
            } catch (_: IOException) {
                // ByteArrayOutputStream never throws.
            }
            return out.toByteArray()
        }
    }
}
